### PR TITLE
feat: add field allocateLoadBalancerNodePorts to KubernetesServiceSpec

### LINKED ISSUE #1772: Allow custom (-options in the) Service object - allocateLoadBalancerNodePorts: false
*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

Allow for custom `Service` (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#servicespec-v1-core) objects to be provided in `KubernetesServiceSpec` (https://gateway.envoyproxy.io/latest/api/config_types.html#kubernetesservicespec) instead of only `ServiceType`.
The only aspect overridden by Envoy Gateway would need be the `selector` object in the `Service`.
Having the Service object exposed to configuration allows for all sorts of custom deployments .

Specifically, I am looking for a way to support `allocateLoadBalancerNodePorts: false`.
Likely others will also start to look into custom: `loadBalancerClass` or other options.

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

See: https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation

### PR DESCRIPTION
**Pull request by den3tsou**
**What this PR does / why we need it**:
This PR adds field `allocateLoadBalancerNodePorts` to [KubernetesServiceSpec](https://gateway.envoyproxy.io/latest/api/config_types.html#kubernetesservicespec) for Envoy Proxy kind. The value will get propagated to the Envoy Proxy Sevice configuration, so users can choose to not use automatically assigned `NodePorts` by setting `false` for this field. Please see [ServiceSpec v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#servicespec-v1-core) for more info.

**Which issue(s) this PR fixes**:
Fixes #1772

### CODE REVIEW

File: api/config/v1alpha1/validation/validate_test.go [hunk 1]
```diff
@@ -602,3 +644,7 @@ func TestGetEnvoyProxyComponentLevelArgs(t *testing.T) {
 		})
 	}
 }
+
+func toPointer[T any](t T) *T {
```
Comment (den3tsou) [inline api/config/v1alpha1/validation/validate_test.go:648]: Is there any good place for this kind of help function?

File: api/config/v1alpha1/validation/validate.go [hunk 2]
```diff
@@ -93,6 +96,18 @@ func validateServiceType(spec *egcfgv1a1.EnvoyProxySpec) []error {
 	return errs
 }
 
+func validateServiceAllocateLoadBalancerNodePorts(spec *egcfgv1a1.EnvoyProxySpec) error {
```
Comment (arkodg) [inline api/config/v1alpha1/validation/validate.go:99]: thoughts on flattening this and above `validateServiceType` into a single function called `validateService` ?

File: api/config/v1alpha1/validation/validate_test.go (hunk 1 above)
Comment (arkodg) [reply api/config/v1alpha1/validation/validate_test.go:648]: yah you can use https://github.com/envoyproxy/gateway/blob/main/internal/utils/ptr/ptr.go

File: internal/infrastructure/kubernetes/resource/resource_test.go [hunk 3]
```diff
@@ -51,6 +52,19 @@ func TestExpectedServiceSpec(t *testing.T) {
 				ExternalTrafficPolicy: corev1.ServiceExternalTrafficPolicyTypeLocal,
 			},
 		},
+		{
+			name: "LoadBalancerWithAllocateLoadBalancerNodePorts",
+			args: args{service: &egcfgv1a1.KubernetesServiceSpec{
+				Type:                          egcfgv1a1.GetKubernetesServiceType(egcfgv1a1.ServiceTypeLoadBalancer),
+				AllocateLoadBalancerNodePorts: &allocateLoadBalancerNodePorts,
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/resource/resource_test.go:59]: can we use `ptrTo(true)` style in this file ?

File: api/config/v1alpha1/validation/validate.go (hunk 2 above)
Comment (den3tsou) [reply api/config/v1alpha1/validation/validate.go:99]: Sounds good to me. I will update it

File: internal/infrastructure/kubernetes/resource/resource_test.go (hunk 3 above)
Comment (den3tsou) [reply internal/infrastructure/kubernetes/resource/resource_test.go:59]: Sounds great. Updated.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**